File: app/src/main/java/com/pnuema/bible/ui/dialogs/VersionSelectionDialog.java
package com.pnuema.bible.ui.dialogs;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.pnuema.bible.R;
import com.pnuema.bible.data.IVersion;
import com.pnuema.bible.data.firefly.Versions;
import com.pnuema.bible.retrievers.BaseRetriever;
import com.pnuema.bible.retrievers.FireflyRetriever;
import com.pnuema.bible.statics.CurrentSelected;
import com.pnuema.bible.statics.LanguageUtils;
import com.pnuema.bible.ui.adapters.VersionSelectionRecyclerViewAdapter;

import java.util.ArrayList;
import java.util.List;
import java.util.Observable;
import java.util.Observer;

public class VersionSelectionDialog extends DialogFragment implements VersionSelectionListener, Observer {
    private NotifyVersionSelectionCompleted mListener;
    private VersionSelectionRecyclerViewAdapter mAdapter;
    private final List<IVersion> mVersions = new ArrayList<>();
    private BaseRetriever mRetriever = new FireflyRetriever();

    @Override
    public void onVersionSelected(final String version) {
        CurrentSelected.setVersion(version);
        if (mListener != null) {
            mListener.onSelectionComplete(version);
        }
        dismiss();
    }

    public static VersionSelectionDialog instantiate(final NotifyVersionSelectionCompleted notifySelectionCompleted) {
        final VersionSelectionDialog dialog = new VersionSelectionDialog();
        dialog.setListener(notifySelectionCompleted);

        return dialog;
    }

    private void setListener(final NotifyVersionSelectionCompleted listener) {
        mListener = listener;
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.dialog_version_picker, container);

        mAdapter = new VersionSelectionRecyclerViewAdapter(mVersions, this);
        final RecyclerView recyclerView = view.findViewById(R.id.versionRecyclerView);
        recyclerView.setAdapter(mAdapter);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        mRetriever.addObserver(this);
        mRetriever.getVersions();
    }

    @Override
    public void onPause() {
        super.onPause();
        mRetriever.deleteObservers();
    }

    @Override
    public void update(final Observable observable, final Object o) {
        mVersions.clear();
        if (o instanceof Versions) {
            final String lang = LanguageUtils.getISOLanguage();
            for (final IVersion version : ((Versions)o).getVersions()) {
                if (version.getLanguage().contains(lang)) {
                    mVersions.add(version);
                }
            }
            mAdapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/pnuema/android/bible/ui/dialogs/VersionSelectionDialog.kt
package com.pnuema.android.bible.ui.dialogs

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.RecyclerView
import com.pnuema.android.bible.R
import com.pnuema.android.bible.data.IVersion
import com.pnuema.android.bible.data.firefly.Versions
import com.pnuema.android.bible.retrievers.FireflyRetriever
import com.pnuema.android.bible.statics.CurrentSelected
import com.pnuema.android.bible.statics.LanguageUtils
import com.pnuema.android.bible.ui.adapters.VersionSelectionRecyclerViewAdapter
import java.util.*

class VersionSelectionDialog(notifySelectionCompleted: NotifyVersionSelectionCompleted) : DialogFragment(), VersionSelectionListener {
    private var mListener: NotifyVersionSelectionCompleted = notifySelectionCompleted
    private lateinit var mAdapter: VersionSelectionRecyclerViewAdapter
    private val mVersions = ArrayList<IVersion>()
    private val mRetriever = FireflyRetriever()

    companion object {
        fun instantiate(notifySelectionCompleted: NotifyVersionSelectionCompleted): VersionSelectionDialog {
            return VersionSelectionDialog(notifySelectionCompleted)
        }
    }

    override fun onVersionSelected(version: String) {
        CurrentSelected.setVersion(version)
        mListener.onSelectionComplete(version)
        dismiss()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.dialog_version_picker, container)

        mAdapter = VersionSelectionRecyclerViewAdapter(mVersions, this)
        val recyclerView = view.findViewById<RecyclerView>(R.id.versionRecyclerView)
        recyclerView.adapter = mAdapter

        return view
    }

    override fun onResume() {
        super.onResume()
        mRetriever.getVersions().observe(viewLifecycleOwner, androidx.lifecycle.Observer { versions ->
            mVersions.clear()
            if (versions is Versions) {
                val lang = LanguageUtils.getISOLanguage()
                for (version in versions.versions) {
                    if (version.language.contains(lang)) {
                        mVersions.add(version)
                    }
                }
                mAdapter.notifyDataSetChanged()
            }
        })
    }
}